crosstalk selection

casperhart opened the pull request:
- display_scatter modify class for crosstalk
- display_scatter change behaviour of selection to highlight rather than colour
- display_scatter support crosstalk selection
- animate_tour add width and height parameters
- display_scatter add button for colouring selected points

TODO: support crosstalk filtering.

The behaviour for colouring points is now two steps. I.e. points are first selected using the selection tool and then coloured using the new 'brush' button. This is so that the selection can be reflected in linked plots without needing to colour.

Tested using the following code:

```r
# remotes::install_github("jcheng5/d3scatter")
library(d3tourr)
library(crosstalk)
library(d3scatter)

shared_flea <- SharedData$new(tourr::flea)

a <- d3scatter(shared_flea, ~tars1, ~head, ~species, width = "100%")
b <- animate_tour(
  shared_flea,
  display = display_scatter(tour_aes(colour = species)),
  render_opts =  list(max_bases = 10), tour_path = tourr::grand_tour(2),
  width = "100%"
)

bscols(
  a, b
)
```

earowang commented:
Wow, you're on 🔥 

Once we select the colour from the colour tool and hit the "OK" button, can these points be coloured, instead of hit the "brush" button to confirm again?

earowang commented:
Do you think if the "reset" button should also reset colours to the originals?

casperhart commented:
> Do you think if the "reset" button should also reset colours to the originals?

Yes, I'll add this in soon.

> Once we select the colour from the colour tool and hit the "OK" button, can these points be coloured, instead of hit the "brush" button to confirm again?

The colour selection is browser-dependent, and not all have an "OK" button (or any other button). My concern with automatically applying the colour is that if no points are selected, then the colour of every point will be affected which may not be desired. If it is desired behaviour, it should be pretty easy to implement though.

earowang commented:
I find it redundant to have the "brush" button to double confirm. But I don't understand what you mean by "the colour of every point will be affected" if no points are selected.

earowang commented:
The other issue is that we can enable persistent selections with {plotly} by holding the "shift" key. But {d3tourr} swipes the previous selection away. 

```r
library(d3tourr)
library(crosstalk)
library(plotly)

shared_flea <- SharedData$new(tourr::flea)

gg <- ggplot(shared_flea, aes(x = tars1, y = head)) +
  geom_point()

a <- ggplotly(gg) %>% 
  highlight(on = "plotly_click", dynamic = TRUE)
b <- animate_tour(
  shared_flea,
  display = display_scatter(tour_aes(colour = species)),
  render_opts =  list(max_bases = 10), tour_path = tourr::grand_tour(2),
  width = "100%"
)

bscols(
  a, b
)
```

![image](https://user-images.githubusercontent.com/2456216/143797339-638b2cd2-cee6-4c30-bbe7-ef97b5a01c63.png)

casperhart commented:
> The other issue is that we can enable persistent selections with {plotly} by holding the "shift" key. But {d3tourr} swipes the previous selection away.

This is a bug in plotly. In the crosstalk event which is broadcasted from the plotly widget, it only includes the latest selection when it should include the entire selection.

E.g. in the following example selecting one point at a time, the `value` field in the console should contain all selected points, rather than just a single point.
![image](https://user-images.githubusercontent.com/39182232/143826419-4297ffd2-80ec-41bd-82af-357e1dee5c93.png)

https://rstudio.github.io/crosstalk/authoring.html#:~:text=Any%20data%20point%20whose%20key%20does%20not%20appear%20in%20the%20string%20array%20should%20be%20considered%20deselected.

casperhart commented:
> I find it redundant to have the "brush" button to double confirm. But I don't understand what you mean by "the colour of every point will be affected" if no points are selected.

If we make the colour selection effective immediately, then if decide to choose a colour before selecting any points, the colour of _all_ the points will change. 

Plus, in most browsers there is no "Okay" button or any other confirmation on the colour selector, so in RStudio the colour would probably be applied even before you click okay. 

However if the reset button also resets the colour, then this won't really be an issue so I'll have a go at implementing it later in the week.

casperhart commented:
I have updated this so that the colour selection is applied immediately on change, i.e. as soon as the selection window is closed.
I have left the brush button with the same functionality, so that the current colour can be applied without needing to open the colour selector.

earowang commented:
{plotly} persistent selection works well with {DT} as expected. Perhaps you can look into {DT} source code for implementation? 

```r
library(crosstalk)
library(plotly)

shared_flea <- SharedData$new(tourr::flea)

gg <- ggplot(shared_flea, aes(x = tars1, y = head)) +
  geom_point()

a <- ggplotly(gg) %>% 
  highlight(on = "plotly_click", dynamic = TRUE)
library(DT)
c <- datatable(shared_flea)

bscols(
  a, c
)
```

earowang commented:
In the meanwhile, should the "reset" button also reset the transparency of points to the originals?

casperhart commented:
I've implemented the persistent selection similar to {DT}. I also modified the reset behaviour to also reset the current selection and revert to orbit controls. The current time / frame and selected colour are not affected by the reset.

earowang reviewed (approved):
Why are these S3 methods removed? coz of {crosstalk}?

earowang commented on DESCRIPTION:
```diff
@@ -20,7 +20,8 @@ Imports:
     grDevices,
-    jsonlite
+    jsonlite,
+    crosstalk
```
Oh, I'd suggest to list {crosstalk} as `Suggests` instead of `Imports`, since it's optional.

earowang commented on R/animate_tour.R:
```diff
@@ -35,9 +36,21 @@ animate_tour <- function(data,
+                         height = NULL) {
   col_spec <- rlang::enquo(cols)
+
+  if (crosstalk::is.SharedData(data)) {
```
```suggestion
  if (inherits(data, "SharedData")) {
    rlang::is_installed("crosstalk")
```

earowang commented on R/animate_tour.R:
```diff
@@ -82,39 +95,29 @@ animate_tour <- function(data,
-#' @export
-str.tour_animation <- function(object, ...) {
-  jsonlite::toJSON(object$plot_data, digits = 4, auto_unbox = TRUE)
+    dependencies = crosstalk::crosstalkLibs(),
```
if else for dependencies.
